activetable/activetable.py

```diff
@@ -0,0 +1,297 @@
+    answers_correct = Dict(scope=Scope.user_state, default=None)
+    # The number of points awarded.
+    score = Float(scope=Scope.user_state)
+    # The number of attempts used.
+    attempts = Integer(scope=Scope.user_state, default=0)
+
+    has_score = True
+
+    @property
+    def num_correct_answers(self):
+        """The number of correct answers during the last check."""
+        if self.answers_correct is None:
+            return None
+        return sum(self.answers_correct.itervalues())
+
+    @property
+    def num_total_answers(self):
+        """The total number of answers during the last check."""
+        if self.answers_correct is None:
+            return None
+        return len(self.answers_correct)
+
+    def parse_fields(self):
+        """Parse the user-provided fields into more processing-friendly structured data."""
+        if self.content:
+            self.thead, self.tbody = parse_table(self.content)
+        else:
+            self.thead = self.tbody = None
+            return
+        if self.column_widths:
+            self._column_widths = parse_number_list(self.column_widths)
+        else:
+            self._column_widths = [800 / len(self.thead)] * len(self.thead)
+        if self.row_heights:
+            self._row_heights = parse_number_list(self.row_heights)
+        else:
+            self._row_heights = [36] * (len(self.tbody) + 1)
+
+    def postprocess_table(self):
+        """Augment the parsed table definition with further information.
+
+        The additional information is taken from other content and student state fields.
+        """
+        self.response_cells = {}
+        for row, height in zip(self.tbody, self._row_heights[1:]):
+            row['height'] = height
+            if row['index'] & 1:
+                row['class'] = 'even'
+            else:
+                row['class'] = 'odd'
+            for cell in row['cells']:
+                cell.id = 'cell_{}_{}'.format(row['index'], cell.index)
+                cell.classes = ''
+                if not cell.is_static:
+                    self.response_cells[cell.id] = cell
+                    cell.classes = 'active'
+                    cell.value = self.answers.get(cell.id)
+                    cell.height = height - 2
+                    if isinstance(cell, NumericCell) and cell.abs_tolerance is None:
+                        cell.set_tolerance(self.default_tolerance)
+
+    def get_status(self):
+        """Status dictionary passed to the frontend code."""
+        return dict(
+            answers_correct=self.answers_correct,
+            num_correct_answers=self.num_correct_answers,
+            num_total_answers=self.num_total_answers,
+            score=self.score,
+            max_score=self.max_score,
+            attempts=self.attempts,
+            max_attempts=self.max_attempts,
+        )
+
+    def student_view(self, context=None):
+        """Render the table."""
+        self.parse_fields()
```

Comment:
I'm surprised to see the fields being parsed each time.  Not sure why that bothers me, but it seems like significant work that could be done just once.